File: lattice.core/Core/CallStack.vb
Imports Humanizer
Imports ObjectIR.Core.AST

Namespace Core
    Public Class CallStack
        Public Property Method As MethodNode
        Public Property IP As Integer
        Public Property Locals As Dictionary(Of String, Object)
        Public Property Args As Dictionary(Of String, Object)
        Public Property This As ManagedObject
        Public Property Previous As CallStack
        Public Property EvaluationStack As Stack(Of Object)

        Public Sub New(method As MethodNode, Optional thisObj As ManagedObject = Nothing)
            Me.Method = method
            Me.IP = 0
            Me.Locals = New Dictionary(Of String, Object)()
            Me.Args = New Dictionary(Of String, Object)()
            Me.This = thisObj
            Me.Previous = Nothing
            Me.EvaluationStack = New Stack(Of Object)()
        End Sub

        Public Function PushFrame(newMethod As MethodNode, Optional thisObj As ManagedObject = Nothing) As CallStack
            Dim frame As New CallStack(newMethod, thisObj)
            frame.Previous = Me
            Return frame
        End Function

        Public Function PopFrame() As CallStack
            Return Me.Previous
        End Function

        Public Overrides Function ToString() As String
            Dim name = If(Method IsNot Nothing AndAlso Method.Name IsNot Nothing, Method.Name, "unknown")
            Return $"at {name} @ {IP} with args {Me.Args.Humanize()}"
            'Return GetStackTrace()
        End Function

        Public Function GetStackTrace() As String
            Dim sb As New System.Text.StringBuilder()
            Dim current As CallStack = Me
            While current IsNot Nothing
                sb.AppendLine(current.ToString())
                current = current.Previous
            End While
            Return sb.ToString()
        End Function
    End Class
End Namespace

File: Lattice.Runtime/Runtime/CallStack.cs
using ObjectIR.Core.AST;

namespace lattice.Core;

public class CallStack
{
    public MethodNode Method { get; set; }
    public int IP { get; set; }
    public Dictionary<string, object> Locals { get; set; } = new();
    public Dictionary<string, object> Args { get; set; } = new();
    public ManagedObject? This { get; set; }
    public CallStack? Previous { get; set; }
    public Stack<object> EvaluationStack { get; set; } = new();
    public bool BreakRequested { get; set; }
    public bool ContinueRequested { get; set; }

    public CallStack(MethodNode method, ManagedObject? thisObj = null)
    {
        Method = method;
        IP = 0;
        Locals = new Dictionary<string, object>();
        Args = new Dictionary<string, object>();
        This = thisObj;
        if (thisObj != null)
        {
            Args["this"] = thisObj;
        }
        Previous = null;
        EvaluationStack = new Stack<object>();
    }

    public CallStack PushFrame(MethodNode newMethod, ManagedObject? thisObj = null)
    {
        var frame = new CallStack(newMethod, thisObj);
        frame.Previous = this;
        return frame;
    }

    public CallStack? PopFrame() => Previous;

    public override string ToString()
    {
        var name = Method?.Name ?? "unknown";
        return $"at {name} @ {IP}";
    }

    public string GetStackTrace()
    {
        var sb = new System.Text.StringBuilder();
        var current = this;
        while (current != null)
        {
            sb.AppendLine(current.ToString());
            current = current.Previous;
        }
        return sb.ToString();
    }
}
